schema_registry.cc

```diff
@@ -150,6 +152,18 @@ schema_ptr schema_registry::get_or_load(table_schema_version v, const schema_loa
 schema_ptr schema_registry_entry::load(frozen_schema fs) {
     _frozen_schema = std::move(fs);
     auto s = get_schema();
+    // If this is a view so this schema also needs a reference to the base
+    // table.
+    if (s->is_view()) {
+        if (!s->view_info()->base_info()) {
+            auto& db = service::get_local_storage_proxy().get_db().local();
+            // This line might throw a no_such_column_family
+            // It should be fine since if we tried to register a view for which
+            // we don't know the base table, our registry is broken.
+            schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
+            s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*base_schema));
```

Comment:
I'm not sure this is correct.

There is no guarantee that the current base table schema returned by db.find_schema() will be consistent with the view schema here, since this is a free-floating view schema and may be out of date with the current schema completely.

I wonder how come we haven't run into this problem before.

Reply:
We don't allow changes to the base table that will break it's views and we don't allow altering the actual query of the view.
It can be problematic if the view was already dropped on that node and then the table has changed. We currently don't have any mechanism to bind a view to it's base table's schema version.

Reply:
@tgrabiec , another think that has just came to my mind, this view as you defined it is a free-floating view, it came from 
another node and we don't know anything about it except for it's (immutable) scehma part. This means that this view will not be used for writing since each node is responsible for writing to it's own view but only for reading, when reading, the associated base table schema is used with reads only to determine if a row is shadowed by a shadowable tombstone,
Here is the code that uses it (which actually cased the segfault in #7420):
https://github.com/scylladb/scylla/blob/aa4c359cff89e8a5c3fdc4a396ef1ce973725302/mutation_partition.cc#L1789-L1813

So the condition uses it to to determine if to consider the row marker for shadowable tombstone. Currently it segfault exactly by what you mentioned, this is a free floating view schema, so our options are (with regards to existing versions):
1. not allowing such reads. (by throwing maybe instead of segfaulting)
2.  going for the wishful thinking that it will be ok, this should only be bad in a *schema race condition* where the view is dropped the a table is altered in a way that is problematic to the view and then a node which still hasn't apply the change yet sends read command for the view. (I am not even sure it can happen)

For future versions we can try and add this piece of metadata to the view's schema but we should check if it is even needed (since we don't allow any change to the base table that can ruin the views).

Reply:
To add some infromation:
After reviewing the code, it looks like the condition to create a shadowable tombstone while reading a mutation stream is 
dependant uppon this information (has_base_non_pk_columns_in_view_pk == true). So the question is, what are we doing given that this info is not available to us without doing this wishful act? I suggested alternative approach above (just throw - meaning, fail the read).
As for @tgrabiec question [above](https://github.com/scylladb/scylla/pull/7440#discussion_r505633509)
I replied with hand waving that besides the race condition (which is still hand waving since I can't think of how to reliably reproduce this) we don't allow changing the base table in a way that will make views incompatible, here are two code examples to what I claimed :slightly_smiling_face: :
Not alowing changing to incompatible column types:
https://github.com/scylladb/scylla/blob/aa4c359cff89e8a5c3fdc4a396ef1ce973725302/cql3/statements/alter_table_statement.cc#L236-L256
Not allowing dropping a column:
https://github.com/scylladb/scylla/blob/aa4c359cff89e8a5c3fdc4a396ef1ce973725302/cql3/statements/alter_table_statement.cc#L258-L287

@nyh ccing you so you can reply or maybe correct mistakes in my replies if there are such.
@avikivity ccing you so you can help us decide ( in case we have to decide today).

@slivne ccing you so you know what is going on.

Reply:
@tgrabiec I have suggested here the two solutions I have in mind and tried to answer your correctness concern the best I can, given those answers, and the understanding that this possible data corruption is likely to be there for a long time, what do you think we should do?
We might in addition to this or another change we will choose can think of a more correct fix where a materialized view info in schema also contains it's base table version. But this will probably take a lot longer that we can afford for 4.2

Reply:
Let me say from the start that I always thought that it was a mistake for view operations to require the base table to exist. Most of them DO NOT, and the very few that do need only specific information such as is_view() and has_base_non_pk_columns_in_view_pk() - and we could and should have saved this information inside the view itself.
But, given that we do have these very few view functions that need to know the base, the question becomes - do they just need ANY VERSION of the base schema, or a very specific version?
I agree with Eliran that any version would do. has_base_non_pk_columns_in_view_pk() just needs to know whether all the view's key columns were also key columns in the base - or not. The answer to this question cannot change as versions change, because the base's key columns may not change with ALTER TABLE (that's a hard limitation).

That being said, @eliransin I think there's a bug in view_info::make_base_dependent_view_info() -  I'll write a separate comment about that.

Reply:
@nyh Can you please clarify, are you saying that it's always ok to attach any base schema to the view schema, for any purposes, or only for `has_base_non_pk_columns_in_view_pk()`?

Reply:
What do we need it beyond has_base_non_pk_columns_in_view_pk()? It has unfortunately become hard to understand the answer to this :-( The base_dependent_view_info has a base_schema, and it's used in view.cc for things like `reader.upgrade_schema(base_schema)`, but why does it matter, and if the base_schema we use is the newest one, isn't it good enough already?
But more importantly, the case where we use the "wrong" base version is is the case where we fetch an unknown view schema version, this can only happen during a read request (since you cannot write to a view!). So all the complex view update code doesn't come into play at all, and in the read path, I **think** and has_base_non_pk_columns_in_view_pk is the only remaining case where this is used. But the code has simply become too complicated to reason and deciding that this is really true, so I don't know how to "guarantee" that this is the case...

Reply:
`reader.upgrade_schema(base_schema)` is needed so that fragments produced by the readers use the same column ids as were recorded in `base_non_pk_columns_in_view_pk`. 

So this `base_dependent_view_info` is mostly used when updating the view, and we don't really need the base schema to be always attached. We could initialize the mapping lazily, and maybe even not store it in the view_info object. I guess it still makes sense to cache it for performance reasons though, so that we don't compute it every time.

Then the only thing which needs to be always valid is this `has_base_non_pk_columns_in_view_pk` flag, which ideally would be part of the view schema mutations.